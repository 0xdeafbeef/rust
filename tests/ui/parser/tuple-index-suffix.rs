//! See #60210.
//!
//! Check that we hard error on invalid suffixes in tuple indexing subexpressions and struct numeral
//! field names, modulo carve-outs for `{i,u}{32,usize}` at warning level to mitigate ecosystem
//! impact.

struct X(i32,i32,i32);

fn main() {
    let tup_struct = X(1, 2, 3);
    let invalid_tup_struct_suffix = tup_struct.0suffix;
    //~^ ERROR suffixes on a tuple index are invalid
    let carve_out_tup_struct_suffix = tup_struct.0i32;
    //~^ WARN suffixes on a tuple index are invalid

    let tup = (1, 2, 3);
    let invalid_tup_suffix = tup.0suffix;
    //~^ ERROR suffixes on a tuple index are invalid
    let carve_out_tup_suffix = tup.0u32;
    //~^ WARN suffixes on a tuple index are invalid

    numeral_struct_field_name_suffix_invalid();
    numeral_struct_field_name_suffix_carve_out();
}

// Very limited carve outs as a ecosystem impact mitigation implemented in #60186. *Only*
// `{i,u}{32,usize}` suffixes are temporarily accepted.
fn carve_outs() {
    // Ok, only pseudo-FCW warnings.

    let carve_out_i32 = (42,).0i32;     //~ WARN suffixes on a tuple index are invalid
    let carve_out_i32 = (42,).0u32;     //~ WARN suffixes on a tuple index are invalid
    let carve_out_isize = (42,).0isize; //~ WARN suffixes on a tuple index are invalid
    let carve_out_usize = (42,).0usize; //~ WARN suffixes on a tuple index are invalid

    // Not part of the carve outs!
    let error_i8 = (42,).0i8;      //~ ERROR suffixes on a tuple index are invalid
    let error_u8 = (42,).0u8;      //~ ERROR suffixes on a tuple index are invalid
    let error_i16 = (42,).0i16;    //~ ERROR suffixes on a tuple index are invalid
    let error_u16 = (42,).0u16;    //~ ERROR suffixes on a tuple index are invalid
    let error_i64 = (42,).0i64;    //~ ERROR suffixes on a tuple index are invalid
    let error_u64 = (42,).0u64;    //~ ERROR suffixes on a tuple index are invalid
    let error_i128 = (42,).0i128;  //~ ERROR suffixes on a tuple index are invalid
    let error_u128 = (42,).0u128;  //~ ERROR suffixes on a tuple index are invalid
}

fn numeral_struct_field_name_suffix_invalid() {
    let invalid_struct_name = X { 0suffix: 0, 1: 1, 2: 2 };
    //~^ ERROR suffixes on a tuple index are invalid
    match invalid_struct_name {
        X { 0suffix: _, .. } => {}
        //~^ ERROR suffixes on a tuple index are invalid
    }
}

fn numeral_struct_field_name_suffix_carve_out() {
    let carve_out_struct_name = X { 0u32: 0, 1: 1, 2: 2 };
    //~^ WARN suffixes on a tuple index are invalid
    match carve_out_struct_name {
        X { 0u32: _, .. } => {}
        //~^ WARN suffixes on a tuple index are invalid
    }
}

// Unfortunately, it turns out `std::mem::offset_of!` uses the same expect suffix code path.
fn offset_of_suffix() {
    #[repr(C)]
    pub struct Struct<T>(u8, T);

    // Carve outs
    assert_eq!(std::mem::offset_of!(Struct<u32>, 0usize), 0);
    //~^ WARN suffixes on a tuple index are invalid

    // Not part of carve outs
    assert_eq!(std::mem::offset_of!(Struct<u32>, 0u8), 0);
    //~^ ERROR suffixes on a tuple index are invalid
}
