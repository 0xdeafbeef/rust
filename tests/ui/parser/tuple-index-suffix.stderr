error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:11:48
   |
LL |     let invalid_tup_struct_suffix = tup_struct.0suffix;
   |                                                ^^^^^^^ invalid suffix `suffix`

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:13:50
   |
LL |     let carve_out_tup_struct_suffix = tup_struct.0i32;
   |                                                  ^^^^ invalid suffix `i32`
   |
   = help: `i32` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:17:34
   |
LL |     let invalid_tup_suffix = tup.0suffix;
   |                                  ^^^^^^^ invalid suffix `suffix`

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:19:36
   |
LL |     let carve_out_tup_suffix = tup.0u32;
   |                                    ^^^^ invalid suffix `u32`
   |
   = help: `u32` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:31:31
   |
LL |     let carve_out_i32 = (42,).0i32;
   |                               ^^^^ invalid suffix `i32`
   |
   = help: `i32` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:32:31
   |
LL |     let carve_out_i32 = (42,).0u32;
   |                               ^^^^ invalid suffix `u32`
   |
   = help: `u32` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:33:33
   |
LL |     let carve_out_isize = (42,).0isize;
   |                                 ^^^^^^ invalid suffix `isize`
   |
   = help: `isize` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:34:33
   |
LL |     let carve_out_usize = (42,).0usize;
   |                                 ^^^^^^ invalid suffix `usize`
   |
   = help: `usize` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:37:26
   |
LL |     let error_i8 = (42,).0i8;
   |                          ^^^ invalid suffix `i8`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:38:26
   |
LL |     let error_u8 = (42,).0u8;
   |                          ^^^ invalid suffix `u8`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:39:27
   |
LL |     let error_i16 = (42,).0i16;
   |                           ^^^^ invalid suffix `i16`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:40:27
   |
LL |     let error_u16 = (42,).0u16;
   |                           ^^^^ invalid suffix `u16`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:41:27
   |
LL |     let error_i64 = (42,).0i64;
   |                           ^^^^ invalid suffix `i64`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:42:27
   |
LL |     let error_u64 = (42,).0u64;
   |                           ^^^^ invalid suffix `u64`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:43:28
   |
LL |     let error_i128 = (42,).0i128;
   |                            ^^^^^ invalid suffix `i128`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:44:28
   |
LL |     let error_u128 = (42,).0u128;
   |                            ^^^^^ invalid suffix `u128`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:48:35
   |
LL |     let invalid_struct_name = X { 0suffix: 0, 1: 1, 2: 2 };
   |                                   ^^^^^^^ invalid suffix `suffix`

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:51:13
   |
LL |         X { 0suffix: _, .. } => {}
   |             ^^^^^^^ invalid suffix `suffix`

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:57:37
   |
LL |     let carve_out_struct_name = X { 0u32: 0, 1: 1, 2: 2 };
   |                                     ^^^^ invalid suffix `u32`
   |
   = help: `u32` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:60:13
   |
LL |         X { 0u32: _, .. } => {}
   |             ^^^^ invalid suffix `u32`
   |
   = help: `u32` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

warning: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:71:50
   |
LL |     assert_eq!(std::mem::offset_of!(Struct<u32>, 0usize), 0);
   |                                                  ^^^^^^ invalid suffix `usize`
   |
   = help: `usize` is *temporarily* accepted on tuple index fields as it was incorrectly accepted on stable for a few releases
   = help: on proc macros, you'll want to use `syn::Index::from` or `proc_macro::Literal::*_unsuffixed` for code that will desugar to tuple field access
   = help: see issue #60210 <https://github.com/rust-lang/rust/issues/60210> for more information

error: suffixes on a tuple index are invalid
  --> $DIR/tuple-index-suffix.rs:75:50
   |
LL |     assert_eq!(std::mem::offset_of!(Struct<u32>, 0u8), 0);
   |                                                  ^^^ invalid suffix `u8`

error: aborting due to 13 previous errors; 9 warnings emitted

